server-timing/class-perflab-server-timing.php

```diff
@@ -0,0 +1,265 @@
+					if ( ! current_user_can( $this->registered_metrics_data[ $metric->get_slug() ]['access_cap'] ) ) {
+						return null;
+					}
+
+					return $this->format_metric_header_value( $metric );
+				},
+				$this->registered_metrics
+			),
+			function( $value ) {
+				return null !== $value;
+			}
+		);
+
+		return implode( ', ', $metric_header_values );
+	}
+
+	/**
+	 * Returns whether an output buffer should be used to gather Server-Timing metrics during template rendering.
+	 *
+	 * Without an output buffer, it is only possible to cover metrics from before serving the template, i.e. before
+	 * the HTML output starts. Therefore sites that would like to gather metrics while serving the template should
+	 * enable this via the {@see 'perflab_server_timing_use_output_buffer'} filter.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @return bool True if an output buffer should be used, false otherwise.
+	 */
+	public function use_output_buffer() {
+		/**
+		 * Filters whether an output buffer should be used to be able to gather additional Server-Timing metrics.
+		 *
+		 * Without an output buffer, it is only possible to cover metrics from before serving the template, i.e. before
+		 * the HTML output starts. Therefore sites that would like to gather metrics while serving the template should
+		 * enable this.
+		 *
+		 * @since n.e.x.t
+		 *
+		 * @param bool $use_output_buffer Whether to use an output buffer.
+		 */
+		return apply_filters( 'perflab_server_timing_use_output_buffer', false );
+	}
+
+	/**
+	 * Hook callback for the 'template_include' filter.
+	 *
+	 * This effectively initializes the class to send the Server-Timing header at the right point.
+	 *
+	 * This method is solely intended for internal use within WordPress.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param mixed $passthrough Optional. Filter value. Default null.
+	 * @return mixed Unmodified value of $passthrough.
+	 */
+	public function on_template_include( $passthrough = null ) {
+		if ( ! $this->use_output_buffer() ) {
+			$this->send_header();
+			return $passthrough;
+		}
+
+		ob_start();
+		add_action(
+			'shutdown',
+			function() {
+				$output = ob_get_clean();
+				$this->send_header();
+				echo $output;
+			},
+			// phpcs:ignore PHPCompatibility.Constants.NewConstants
+			defined( 'PHP_INT_MIN' ) ? PHP_INT_MIN : -1000
+		);
+		return $passthrough;
+	}
+
+	/**
+	 * Formats the header segment for a single metric.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param Perflab_Server_Timing_Metric $metric The metric to format.
+	 * @return string|null Segment for the Server-Timing header, or null if no value set.
+	 */
+	private function format_metric_header_value( Perflab_Server_Timing_Metric $metric ) {
+		$value = $metric->get_value();
+
+		// If no value is set, make sure it's just passed through.
+		if ( null === $value ) {
+			return null;
+		}
+
+		if ( is_float( $value ) ) {
+			$value = round( $value, 2 );
+		}
+		return sprintf( 'wp-%1$s;dur=%2$s', $metric->get_slug(), $value );
```

Comment:
Having the `wp-` prefix here means that all metrics will be prefixed. It also means that if someone decide to add their own metric in a 3rd party plugin with a custom name, it will also be prefixed with `wp-` which can be confusing. Maybe we shouldn't hardcode this prefix here and let metrics decide which name to use in the header?

Reply:
We can add filter for it? Also we can do it in next release WDYT?

Reply:
Ok, got it. Makes sense to me. Thanks.